### PR TITLE
Add namespace/docType aliasing

### PR DESCRIPTION
**Pull request by akkomar**
Adds support for namespace/docType aliases (https://github.com/mozilla/gcp-ingestion/issues/424).

Aliasing configuration is stored in json config file, to be loaded from GCS. Decoder should be redeployed on configuration changes, likewise on schema changes.
If configuration file is present, schemas will be aliased in the `SchemaStore`, transparently to the rest of the pipeline.

### CODE REVIEW

File: ingestion-beam/src/test/resources/schemaAliasing/example-aliasing-config.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "aliases": [
+    {
+      "alias": "some-product/baseline",
+      "base": "glean/baseline"
+    },
```
Comment (akkomar) [inline ingestion-beam/src/test/resources/schemaAliasing/example-aliasing-config.json]: @fbertsch how do you like this way of configuring aliases?

Comment (mreid-moz) [reply ingestion-beam/src/test/resources/schemaAliasing/example-aliasing-config.json]: I thought we would want to alias entire namespaces, so we could have something like 
```
{
  "alias": "some-product",
  "base": "glean"
}
```

without having to specify each individual `docType`.  This would mean `some-product/*` would map to `glean/*`.

However, re-reading #424 seems to match the approach here.

An advantage to aliasing an entire namespace would be that when Glean adds a new doc type, all applications would be able to use it without adding more aliases. A disadvantage would be that we still need to manage the single glean schema being used by multiple doctypes in the mozilla-pipeline-schemas repo and build step.

Comment (fbertsch) [reply ingestion-beam/src/test/resources/schemaAliasing/example-aliasing-config.json]: I prefer this approach. The probe-scraper will keep track of default pings and new pings added per-product, so we won't have to take care of that in mozilla-pipeline-schemas, which seems appropriate since we will also keep track of new repositories in the probe-scraper. That way all of the information is contained to one place.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/JSONSchemaStore.java [hunk 2]
```diff
@@ -18,11 +19,21 @@
...
   }
 
-  protected JSONSchemaStore(ValueProvider<String> schemasLocation) {
-    super(schemasLocation);
+  /**
+   * Returns a SchemaStore based on the contents of the archive at schemasLocation
+   * with additional schemas aliased according to configuration.
+   */
+  public static JSONSchemaStore of(ValueProvider<String> schemasLocation,
+      @Nullable ValueProvider<String> aliasingConfigurationLocation) {
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/JSONSchemaStore.java]: I have a slight preference for simplifying this by removing the single argument constructor and making this a required field rather than nullable, since the real use in the Decoder will always be passing in both value providers, even if the produced value ends up being empty.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+package com.mozilla.telemetry.schemas;
+
+import static org.apache.commons.lang3.builder.ToStringStyle.SHORT_PREFIX_STYLE;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+class SchemaAlias {
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java]: Slight preference for making this an AutoValue class to reduce some boilerplate and make this more consistent with AutoValue usage elsewhere in this codebase and in Beam.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAliasingConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+package com.mozilla.telemetry.schemas;
+
+import static org.apache.commons.lang3.builder.ToStringStyle.SHORT_PREFIX_STYLE;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.List;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+class SchemaAliasingConfiguration {
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAliasingConfiguration.java]: I'd also be interested in this being AutoValue. Unless hooking up an AutoValue class for Jackson is a pain.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java [hunk 5]
```diff
@@ -97,13 +105,15 @@ int numLoadedSchemas() {
...
+    Multimap<String, String> schemaAliasesMap;
     InputStream inputStream;
     try {
+      schemaAliasesMap = loadSchemaAliasesMap();
       Metadata metadata = FileSystems.matchSingleFileSpec(schemasLocation.get());
       ReadableByteChannel channel = FileSystems.open(metadata.resourceId());
       inputStream = Channels.newInputStream(channel);
     } catch (IOException e) {
-      throw new IOException("Exception thrown while fetching from configured schemasLocation", e);
+      throw new IOException("Exception thrown while loading schemas configuration", e);
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java]: I'd prefer to keep schemaAliasesLocation as separated as possible from schemasLocation rather than nesting the loading of aliases inside loadAllSchemas.

Can we provide a loadSchemaAliases() method that gets called first and then either pass in the value to this method or have that method set a private variable to store the aliases?

That will separate concerns a bit more and allow us to keep these exception messages as specific as possible.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParsePayloadTest.java [hunk 6]
```diff
@@ -36,7 +36,7 @@ public void testOutput() {
         .apply("AddAttributes", MapElements.into(TypeDescriptor.of(PubsubMessage.class))
             .via(element -> new PubsubMessage(element.getPayload(), ImmutableMap.of(
                 "document_namespace", "test", "document_type", "test", "document_version", "1"))))
-        .apply(ParsePayload.of(schemasLocation));
+        .apply(ParsePayload.of(schemasLocation, null));
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParsePayloadTest.java]: instead of nulls, I'd prefer we pass in a value provider via `pipeline.newProvider` with a null value, or if that doesn't work a `StaticValueProvider.of(null)`.

**Comment by akkomar** [comment]
> Are the glean schemas distinctive enough that we can simply inspect the contents of a JSON payload and determine if it's a glean ping, and further infer which type of glean ping it is?

Interesting idea. Yes - I think that what `$schema` keyword is for, and we [should](https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/dev/validation/glean/baseline.1.all.pass.json#L2) be sending it.
`$schema` seems to be equivalent to @fbertsch's suggestion [here](https://github.com/mozilla/gcp-ingestion/issues/424#issue-408319725):
> We could additionally consider adding the aliased namespace and docType as metadata to the ping.


This has the benefit of making adding aliasing easier for clients other than Glean-based (in the current approach we'd need to generalize "aliasing pipeline" for them) at the expense of complicating decoder a little.
I don't see why this wouldn't work for Glean. I also don't see any huge drawbacks from the whole pipeline point of view (apart from complicating schema validation contract, but that's what aliasing does anyway). /cc @fbertsch @mreid-moz

**Comment by jklukas** [comment]
I totally missed that @fbertsch already suggested adding schema to the ping. I wish I had realized the implications earlier in the discussion.

Having a required field in the general Glean schema to identify the ping type sounds like it would give us a robust basis for routing schemas appropriately, or maybe `$schema` already gives us what we need? I don't fully understand the implications of the discussion in https://github.com/mozilla/gcp-ingestion/pull/508#discussion_r270333335 above.

Being able to remove a dependency on the pipeline for scraping and publishing schema aliases would be a good thing for stability of the Decoder, in my opinion.

**Comment by fbertsch** [comment]
Deciding which schemas to use at validation-time seems totally reasonable to me. In fact the Glean schemas already have that information: https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/dev/schemas/glean/baseline/baseline.1.schema.json#L7

We could generalize that: `if ping contains "$schema" key then use that schema instead of namespace/doctype`. [In fact, that's what the ref docs indicate the field should be used for.](https://json-schema.org/latest/json-schema-core.html#rfc.section.7). They don't need to be valid URLs, and we could retrieve them from mps if it is prefixed with `moz://`.

**Comment by mreid-moz** [comment]
My reservation is that this gets us back into having to parse the _contents_ of a document before we can determine which schema to apply. As things stand now we can determine it completely from the submission URL, and with the aliasing support, we can do it from a combination of the submission URL and the aliases list.

This also effectively changes the meaning of the `docType` in the submission URL to "target table", and leaves it to the contents of the ping itself to tell us its `docType`. This is not necessarily a bad thing, but it is a pretty fundamental change. Would we do the same for `docVersion`?

I agree that it would be nice to simplify the logic and tracking of this information, but I worry about pushing that logic inside the ping itself.

**Comment by jklukas** [comment]
> This also effectively changes the meaning of the docType in the submission URL to "target table", and leaves it to the contents of the ping itself to tell us its docType

> I agree that it would be nice to simplify the logic and tracking of this information, but I worry about pushing that logic inside the ping itself.

To address those two concerns, maybe we could introduce a new concept of "scope" to namespace. What if we had the SomeProduct browser specify namespace as something like `glean.some-product` rather than as `some-product`:

    https://incoming.t.m.o/submit/glean.some-product/baseline/1/{uuid}
    https://incoming.t.m.o/submit/glean+some-product/baseline/1/{uuid}

If there's no `.` in the namespace, then we assume the "standard" scope and apply the given namespace to determine both the validation schema and the destination table. If a `.` appears then the validation scope namespace is `glean` and the destination table namespace is `some-product`.

**Comment by relud** [comment]
if it's not too late to use `glean+` or `glean.` or even `https://incoming.t.m.o/glean/some-product/baseline/1/{uuid}`, i think that's an excellent solution

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java (hunk 3 above)
Comment (akkomar) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java]: Nice, I didn't know about AutoValue.
I thought about Lombok when generating this, but didn't want to introduce this library here.

**Comment by mreid-moz** [comment]
How does everyone feel about taking this PR + implementation as a functional first step, then following up with alternate design / architecture proposal(s) separately?

It seems to address the immediate problem of handling reference browser and fenix schemas without a ton of manual schema-related work, and if we update the alias config "by hand" in the short term, the automation around keeping the mapping file up to date could wait until we've had a chance to have a round of design discussions.

**Comment by akkomar** [comment]
> How does everyone feel about taking this PR + implementation as a functional first step, then following up with alternate design / architecture proposal(s) separately?
> 
I am for it - we need a solution sooner than later. Btw., regarding automation - Frank mentioned that there will be some Glean config generation anyway related to table schemas, so having "Glean ping pipeline" shouldn't be something we're tyring to avoind here.

Out of the ideas mentioned here I like Jeff's idea:
> What if we had the SomeProduct browser specify namespace as something like glean.some-product or glean+some-product rather than as some-product

the most. But since it involves changing public api's contract (so also requires minor changes in Glean client library) I'd argue to not introduce it hastily.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java (hunk 5 above)
Comment (akkomar) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java]: I went one step further and decoupled loading aliases from loading schemas. This way it's not mixed with archive-reading code and would be easy to change/remove in the future.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java [hunk 7]
```diff
@@ -31,8 +31,9 @@
  */
 public class ParsePayload extends MapElementsWithErrors.ToPubsubMessageFrom<PubsubMessage> {
 
-  public static ParsePayload of(ValueProvider<String> schemasLocation) {
-    return new ParsePayload(schemasLocation);
+  public static ParsePayload of(ValueProvider<String> schemasLocation,
+      ValueProvider<String> aliasingConfigurationLocation) {
+    return new ParsePayload(schemasLocation, aliasingConfigurationLocation);
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java]: I'd prefer the variable names follow the option names all the way through, so I'd like to see `aliasingConfigurationLocation` change to `schemaAliasesLocation` throughout this PR.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+abstract class SchemaAlias {
+
+  @JsonCreator
+  static SchemaAlias create(@JsonProperty(value = "base", required = true) String base,
+      @JsonProperty(value = "alias", required = true) String alias) {
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java]: The verbosity needed for the jackson annotations is annoying and kind of ruins the brevity benefit of AutoValue classes, but I still thank you for making this change as AutoValue offers nice protection from some of the subtle bugs that can creep in when managing fields by hand.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/integration/StorageIntegrationTest.java [hunk 9]
```diff
@@ -125,12 +125,16 @@ public void testDecoderWithRemoteInputAndOutput() throws Exception {
   public void testCompileDataflowTemplate() throws Exception {
     String gcsPath = "gs://" + bucket;
 
+    String aliases = System.getProperty("user.dir")
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/integration/StorageIntegrationTest.java]: We don't build absolute paths with `"user.dir"` elsewhere in the tests; is there any reason we can't simply pass the relative path?

If we want to introduce more robust path handling for tests, I'd prefer that be a separate PR.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java (hunk 8 above)
Comment (akkomar) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaAlias.java]: Still, that's a little bit less boilerplate.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/integration/StorageIntegrationTest.java (hunk 9 above)
Comment (akkomar) [reply ingestion-beam/src/test/java/com/mozilla/telemetry/integration/StorageIntegrationTest.java]: Thanks and sorry, that's probably leftover from some experiments.

**Comment by whd** [comment]
> How does everyone feel about taking this PR + implementation as a functional first step, then following up with alternate design / architecture proposal(s) separately?

+1. I haven't been following this discussion but probably should have been, so a follow up architecture proposal sounds swell.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java [hunk 10]
```diff
@@ -131,10 +140,66 @@ private void loadAllSchemas() throws IOException {
...
+          String docTypeWithVersionAlias = docTypeAlias + versionAndSuffix;
+
+          tempSchemas.put(namespaceWithDocTypeAlias + "/" + docTypeWithVersionAlias, schema);
+        }
+      }
+    });
+
+    schemas = tempSchemas;
+    dirs = tempDirs;
+  }
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java]: This is a pretty elegant approach to merging in aliased schemas to the existing schema map leaving the existing interface for looking up schemas unchanged. The one drawback I see to this is that it doesn't give ParsePayload any way of knowing that it was a glean schema being used for validation.

We had discussed previously adding "validation_namespace" and "validation_schema" attributes to the PubsubMessage in the case that an alias was followed to determine a schema to validate against. Those attributes could then be used by the Debug View pipeline to filter out any non-Glean messages. Does that still seem useful? If so, we may need to change the interface here so that we can return the name of the underlying schema returned.

Comment (akkomar) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java]: > We had discussed previously adding "validation_namespace" and "validation_schema" attributes to the PubsubMessage in the case that an alias was followed to determine a schema to validate against.

I'm sorry, I'm not sure I follow. If we go with `incoming.t.m.o/glean+product/glean+doc/1/{uuid}`, these attributes can be added to metadata earlier in the processing chain (say, `ParseUri`?). Then `SchemaStore` wouldn't need to be aware of aliasing and we could just revert its changes I'm introducing here.
Does this make sense?

Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java]: Yes, that makes sense. For the time being, then, the republisher topic for debug pings won't have any metadata that distinguishes glean pings from others and you'll have to look inside the payload to determine if something follows the glean schema. If that's okay for your purposes, then it's fine by me.

Comment (akkomar) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/schemas/SchemaStore.java]: > republisher topic for debug pings won't have any metadata that distinguishes glean pings from others

I think that's OK - I can look at `document_namespace`+`document_type` and validate against a list of products that I'm interested in (I could even reuse aliasing configuration introduced in this PR for that).
